pkg/pi/pi.go

```diff
@@ -3,24 +3,77 @@ package pi
 
 import (
 	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/pkg/errors"
 )
 
 // CheckBinID validates the given binary ID.
-func CheckBinID(id string) error {
-	if len(id) != sha256.Size*2 {
-		return errors.Errorf("invalid length of binary ID; expected %d, got %d", sha256.Size*2, len(id))
+func CheckBinID(binID string) error {
+	if len(binID) != sha256.Size*2 {
+		return errors.Errorf("invalid length of binary ID; expected %d, got %d", sha256.Size*2, len(binID))
 	}
-	if id != strings.ToLower(id) {
-		return errors.Errorf("invalid binary ID; expected lowercase, got %q", id)
+	if binID != strings.ToLower(binID) {
+		return errors.Errorf("invalid binary ID; expected lowercase, got %q", binID)
 	}
 	const hex = "0123456789abcdef"
-	for _, r := range id {
+	for _, r := range binID {
 		if !strings.ContainsRune(hex, r) {
-			return errors.Errorf("invalid rune in binary ID; expected hexadecimal digit, got %q", r)
+			return errors.Errorf("invalid character in binary ID; expected hexadecimal digit, got %q", r)
 		}
 	}
 	return nil
 }
+
+// BinID returns the binary ID corresponding to the given binary contents. The
+// binary ID is the computed SHA256 hashsum of the binary contents in lowercase.
+func BinID(content []byte) string {
+	rawHash := sha256.Sum256(content)
+	binID := hex.EncodeToString(rawHash[:])
+	// Sanity check.
+	if err := CheckBinID(binID); err != nil {
+		panic(fmt.Errorf("discripancy between computed binary ID %q and validity check; %+v", binID, err))
+	}
+	return binID
+}
+
+// CacheDir returns the pippi cache directory.
+func CacheDir() (string, error) {
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+	pippiCacheDir := filepath.Join(cacheDir, "pippi")
+	return pippiCacheDir, nil
+}
+
+// BinDir returns the project directory of the given binary ID.
```

Comment:
@karlek Regarding terminology, should we refer to the `.cache/pippi/{BinID}/` directory as a *project directory* of a given binary ID? Or do we want to reserve *project* for later concepts and use *binary directory* instead?

Reply:
Feels like a project is something greater. As in the example of Diablo 1, there could be multiple binaries used in a project.

Reply:
Yeah, I would agree. Lets reserve the term project for those uses, and revert to using "binary directory" here. I'll send a fix to *master* :)